packages/wdio-runner/src/index.ts

```diff
@@ -29,59 +59,62 @@ export default class Runner extends EventEmitter {
-        const isMultiremote = this.isMultiremote = !Array.isArray(this.configParser.getCapabilities())
+        this._config = this._configParser.getConfig() as ConfigOptions
+        this._config.specFileRetryAttempts = (this._config.specFileRetries || 0) - (retries || 0)
+        logger.setLogLevelsConfig(this._config.logLevels, this._config.logLevel)
+        const isMultiremote = this._isMultiremote = !Array.isArray(this._configParser.getCapabilities())
 
         /**
          * create `browser` stub only if `specFiltering` feature is enabled
          */
         let browser = await this._startSession({
-            ...this.config,
-            _automationProtocol: this.config.automationProtocol,
+            ...this._config,
+            // @ts-ignore used in `/packages/webdriverio/src/protocol-stub.ts`
+            _automationProtocol: this._config.automationProtocol,
             automationProtocol: './protocol-stub'
         }, caps)
 
         /**
          * run `beforeSession` command before framework and browser are initiated
          */
-        initialiseWorkerService(this.config, caps, args.ignoredWorkerServices)
-            .map(this.configParser.addService.bind(this.configParser))
-        await runHook('beforeSession', this.config, this.caps, this.specs)
-
-        this.reporter = new BaseReporter(this.config, this.cid, { ...caps })
+        initialiseWorkerService(
+            this._config as WebdriverIO.Config,
+            caps as unknown as WebDriver.DesiredCapabilities,
+            args.ignoredWorkerServices
+        ).map(this._configParser.addService.bind(this._configParser))
+        await runHook('beforeSession', this._config, this._caps, this._specs)
+
+        this._reporter = new BaseReporter(this._config, this._cid, { ...caps })
         /**
          * initialise framework
          */
-        this.framework = initialisePlugin(this.config.framework, 'framework').default
-        this.framework = await this.framework.init(cid, this.config, specs, caps, this.reporter)
-        process.send({ name: 'testFrameworkInit', content: { cid, caps, specs, hasTests: this.framework.hasTests() } })
-        if (!this.framework.hasTests()) {
+        this._framework = initialisePlugin(this._config.framework as string, 'framework').default as unknown as Framework
+        this._framework = await this._framework.init(cid, this._config, specs, caps, this._reporter)
+        process.send!({ name: 'testFrameworkInit', content: { cid, caps, specs, hasTests: this._framework.hasTests() } })
+        if (!this._framework.hasTests()) {
             return this._shutdown(0, retries)
         }
 
-        browser = await this._initSession(this.config, this.caps, browser)
-        this.inWatchMode = Boolean(this.config.watch)
+        browser = await this._initSession(this._config as SingleConfigOption, this._caps, browser as WebdriverIO.BrowserObject) as WebdriverIO.BrowserObject
```

Comment:
What if you declare `browser` as `WebdriverIO.BrowserObject | null` at line 89 `let browser = await this._startSession({`.

Maybe in such a case, the additional `as` can be removed.